content/How_To/material/Material_plugins.md

```diff
@@ -0,0 +1,183 @@
+
+    getClassName() {
+        return "ColorifyPluginMaterial";
+    }
+
+    getCustomCode(shaderType) {
+        return shaderType === "vertex" ? null : {
+            // this is just like before. Multiply the final shader color by
+            // our color. Note that we have access to all shader variables:
+            // we're effectively inserting a piece of code in the shader code.
+            "CUSTOM_FRAGMENT_BEFORE_FRAGCOLOR": `
+                #ifdef COLORIFY
+                    ${this._varColorName}.rgb *= myColor;
+                #endif
+            `,
+
+            // we can even use regexes to replace arbitrary parts of the code.
+            // if your key starts with '!' it's parsed as a Regex.
+            "!diffuseBase\\+=info\\.diffuse\\*shadow;": `
+                diffuseBase += info.diffuse*shadow;
+                diffuseBase += vec3(0., 0.2, 0.8);
+            `,
+        };
+    }
+}
+```
+
+<Playground id="#P8B91Z#32" title="Material plugin example with uniforms"/>
+
+## Applying a plugin to a single material
+
+Even though your plugins can be enabled/disabled as shown above, sometimes you need to apply a plugin to a single material. In this case, instead of calling `RegisterMaterialPlugin`, you can instantiate the plugin directly with the material that will be modified:
+
+```js
+const myPlugin = new BlackAndWhitePluginMaterial(material);
+```
+
+This is also useful for dynamic loading of plugins.
+
+<Playground id="#22HT5Z#4" title="Material plugin applied to a single material"/>
+
+## Caveats
+
+- The code insertion point names are standard over all materials, but not all of them have all insertion points. In general you can expect to have 
+`CUSTOM_VERTEX_DEFINITIONS`, `CUSTOM_VERTEX_MAIN_BEGIN` and `CUSTOM_VERTEX_MAIN_END` in vertex shaders, and `CUSTOM_FRAGMENT_DEFINITIONS`, `CUSTOM_FRAGMENT_MAIN_BEGIN` and `CUSTOM_FRAGMENT_MAIN_END` in fragment shaders. But other `#defines` might not be present.
+- There's no guarantee that using the shader point names / the regular expressions to update code / the variable names with a material plugin will work across Babylon.js versions. We reserve the possibility to update our shader code in a way that would break backward compatibility with those features if we have to. In particular, the color variable name for `standard.fragment.fx` is `color`, while for `pbrMaterial.fragment.fx` it's `finalColor`. If you are writing a plugin targeting multiple materials take care in your code to use different variable names according to the plugin and keep track of Babylon potentially changing (and breaking!) things.
+- `RegisterMaterialPlugin` only adds the plugin to material instantiated AFTER the registration. So it must be run before you add any meshes or create your materials or they won't have the plugin.
+- You can register multiple plugins to the same material (or the entire scene). The `priority` field controls the order the plugins will be executed if they are all enabled.
```

Comment:
We should add an **Examples** section in the same way we have "Node material examples" for node materials (https://doc.babylonjs.com/divingDeeper/materials/node_material/nodeMaterial#node-material-examples).

Examples you can add in addition to the ones listed in the doc:
* volumetric fog: https://playground.babylonjs.com/#SYQW69#947
* grain (to help with banding - see https://forum.babylonjs.com/t/banding-issue-with-point-light/25836/4): https://playground.babylonjs.com/#IQPBS4#46

Also, I think you can link to the `DetailMap`, `Sheen`, `Anisotropy`, ... plugins in the Babylon.js repo as examples of more involved plugins.